pkg/frontend/querymiddleware/cardinality_test.go

```diff
@@ -0,0 +1,341 @@
+					Query: "up",
+				},
+			},
+			want: fmt.Sprintf("tenant-b:up:%d", daysSinceEpoch),
+		},
+		{
+			name: "range query with large range",
+			args: args{
+				userID: "tenant-b",
+				r: &PrometheusRangeQueryRequest{
+					Start: requestTime.UnixMilli(),
+					// Over 24 hours, should add range part to key
+					End:   requestTime.Add(25 * time.Hour).UnixMilli(),
+					Query: "up",
+				},
+			},
+			want: fmt.Sprintf("tenant-b:up:%d:%d", daysSinceEpoch, 1),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			assert.Equal(t, tt.want, generateCacheKey(tt.args.userID, tt.args.r, 24*time.Hour))
+		})
+	}
+}
+
+func Test_cardinalityEstimation_lookupCardinalityForKey(t *testing.T) {
+	ctx := context.Background()
+	c := cache.NewInstrumentedMockCache()
+
+	presentKey := "tenant-a:up:1234:4321"
+	presentValue := uint64(25)
+	marshaled, err := proto.Marshal(&QueryCardinality{presentValue})
+	require.NoError(t, err)
+	c.Store(ctx, map[string][]byte{cacheHashKey(presentKey): marshaled}, 1*time.Hour)
+
+	expectedFetchCount := 0
+	tests := []struct {
+		cache           cache.Cache
+		name            string
+		key             string
+		wantCardinality uint64
+		wantPresent     bool
+	}{
+		{
+			cache:           nil,
+			name:            "nil cache",
+			key:             presentKey,
+			wantCardinality: 0,
+			wantPresent:     false,
+		},
+		{
+			cache:           c,
+			name:            "cache hit",
+			key:             presentKey,
+			wantCardinality: presentValue,
+			wantPresent:     true,
+		},
+		{
+			cache:           c,
+			name:            "cache miss",
+			key:             "not present",
+			wantCardinality: 0,
+			wantPresent:     false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			estimate, ok := (&cardinalityEstimation{cache: tt.cache}).lookupCardinalityForKey(ctx, tt.key)
+			if tt.cache != nil {
+				expectedFetchCount++
+			}
+			assert.Equal(t, expectedFetchCount, c.CountFetchCalls())
+			assert.Equal(t, tt.wantCardinality, estimate)
+			assert.Equal(t, tt.wantPresent, ok)
+		})
+	}
+}
+
+func Test_cardinalityEstimation_Do(t *testing.T) {
+	const numSeries = uint64(25)
+	request := &PrometheusRangeQueryRequest{
+		Start: parseTimeRFC3339(t, "2023-01-31T09:00:00Z").Unix() * 1000,
+		End:   parseTimeRFC3339(t, "2023-01-31T10:00:00Z").Unix() * 1000,
+		Query: "up",
+	}
+	addSeriesHandler := func(estimate, actual uint64) HandlerFunc {
+		return func(ctx context.Context, request Request) (Response, error) {
+			require.NotNil(t, request.GetHints())
+			require.Equal(t, request.GetHints().EstimatedCardinality, estimate)
+
+			queryStats := stats.FromContext(ctx)
+			queryStats.AddFetchedSeries(actual)
+			return &PrometheusResponse{}, nil
+		}
+	}
+	marshaledEstimate, err := proto.Marshal(&QueryCardinality{numSeries})
+	require.NoError(t, err)
+
+	tests := []struct {
+		name              string
+		orgID             string
```

Comment:
[nit] Regardless the internal legacy name, please call it tenant ID. "Org ID" is a very old naming, we're getting rid of.

```suggestion
		tenantID             string
```